deeplearning4j-scaleout/spark/dl4j-spark-parameterserver/src/main/java/org/deeplearning4j/spark/parameterserver/training/SharedTrainingMaster.java

```diff
@@ -0,0 +1,1044 @@
+
+        if (collectTrainingStats)
+            stats.logFitStart();
+
+        if (storageLevelStreams != null)
+            trainingDataPaths.persist(storageLevelStreams);
+
+        long totalDataSetObjectCount = getTotalDataSetObjectCount(trainingDataPaths);
+
+        doIterationPaths(null, network, trainingDataPaths, 1, 1, dataSetObjectsNumExamples);
+
+        if (collectTrainingStats)
+            stats.logFitEnd((int) totalDataSetObjectCount);
+    }
+
+    protected void executeTrainingPathsMDSHelper(SparkComputationGraph network, JavaRDD<String> trainingMultiDataPaths, int dataSetObjectsNumExamples) {
+        if (numWorkers == null)
+            numWorkers = network.getSparkContext().defaultParallelism();
+
+        if (collectTrainingStats)
+            stats.logFitStart();
+        if (storageLevelStreams != null)
+            trainingMultiDataPaths.persist(storageLevelStreams);
+
+        long totalDataSetObjectCount = getTotalDataSetObjectCount(trainingMultiDataPaths);
+
+        int splitNum = 1;
+
+        doIterationPathsMDS(network, trainingMultiDataPaths, splitNum++, 1, dataSetObjectsNumExamples);
+
+
+        if (collectTrainingStats)
+            stats.logFitEnd((int) totalDataSetObjectCount);
+    }
+
+    protected void prepareNetworkAndStuff(SparkComputationGraph network) {
+        if (numWorkers == null)
+            numWorkers = network.getSparkContext().defaultParallelism();
+
+        // set current box as controller, if field is unset
+        if (voidConfiguration.getControllerAddress() == null)
+            voidConfiguration.setControllerAddress(System.getenv("SPARK_PUBLIC_DNS"));
+
+        if (voidConfiguration.getControllerAddress() == null)
+            throw new DL4JInvalidConfigException("Can't get Spark Master local address. Please specify it manually using VoidConfiguration.setControllerAddress(String) method");
+
+        // first of all, we're instantiating ParameterServer shard here
+        Transport transport = voidConfiguration.getTransportType() == TransportType.ROUTED ? new RoutedTransport() : voidConfiguration.getTransportType() == TransportType.BROADCAST ? new MulticastTransport() : this.transport;
+
+        if (transport == null)
+            throw new DL4JInvalidConfigException("No Transport implementation was defined for this training session!");
+
+        // it's safe to call init on CompGraph, it has isInit check within
+        network.getNetwork().init();
+
+        // this instance will be SilentWorker - it'll accept and apply messages, but won't contribute to training. And we init it only once
+        if (isFirstRun.compareAndSet(false, true)) {
+            trainingDriver = new SilentTrainingDriver(network.getNetwork().params(), network.getNetwork().getOptimizer().getStepFunction());
+            VoidParameterServer.getInstance().init(voidConfiguration, transport, trainingDriver);
+        }
+    }
+
+    protected void prepareNetworkAndStuff(SparkDl4jMultiLayer network) {
```

Comment:
Code duplication - very minimal difference between this and the previous method... Maybe use a method with 2 args, a SparkDl4jMultiLayer and SparkComputationGraph, with one always null?